### PR TITLE
[FLINK-17096] [table] Mini-batch group aggregation doesn't expire state even if state ttl is enabled

### PR DESCRIPTION
**Pull request by lsyldliu**
## What is the purpose of the change

*Support state ttl for Mini-Batch Group Agg using StateTtlConfig and refactor Non-Mini-Batch Group Agg state ttl Strategy using StateTtlConfig instead of Timer*

## Brief change log

  - *Support state ttl for Mini-Batch Group Agg using StateTtlConfig*
  - *Support state ttl for Mini-Batch local、global、Incremental Group Agg using StateTtlConfig*
  - *Refactor Non-Mini-Batch Group Agg state ttl Strategy using StateTtlConfig instead of Timer*

## Verifying this change

This change added tests and can be verified as follows:

  - *Added MiniBatchGroupAggFunctionTest#testMiniBatchGroupAggWithStateTtl test validate that expired state will be cleanup*
  - *Added GroupAggFunctionTest#testGroupAggWithStateTtl test validate that expired state will be cleanup.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (yes / no / don't know)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not documented)

### CODE REVIEW

**Comment by lsyldliu** [comment]
@wuchong In this PR, we will support state ttl for Mini-Batch Group Agg using StateTtlConfig and refactor Non-Mini-Batch Group Agg state ttl Strategy using StateTtlConfig instead of Timer, you can help review the code when you are free, thanks.

File: flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/operators/aggregate/MiniBatchGroupAggFunctionTest.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+import static org.apache.flink.table.runtime.util.StreamRecordUtils.record;
+
+/**
+ * Tests for {@link MiniBatchGroupAggFunction}.
+ */
+public class MiniBatchGroupAggFunctionTest extends GroupAggFunctionTestBase {
+
+	private OneInputStreamOperatorTestHarness<BaseRow, BaseRow> createTestHarness(
+		MiniBatchGroupAggFunction aggFunction
+	) throws Exception {
```
Comment (libenchao) [inline flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/operators/aggregate/MiniBatchGroupAggFunctionTest.java:43]: I've never seen this kind of format in flink, maybe we don't need to make a new line for `) throws Exception {`?

File: flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/operators/aggregate/GroupAggFunctionTestBase.java [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+
+	BaseRowHarnessAssertor assertor = new BaseRowHarnessAssertor(
+		outputType.getFieldTypes(),
+		new GenericRowRecordSortComparator(0, new VarCharType(VarCharType.MAX_LENGTH)));
+
+	static final class SumAndCountAgg implements AggsHandleFunction {
+		long sum;
+		boolean sumIsNull;
+		long count;
+		boolean countIsNull;
```
Comment (libenchao) [inline flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/operators/aggregate/GroupAggFunctionTestBase.java:87]: these fields can be all `private`?

File: flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/operators/aggregate/MiniBatchGroupAggFunctionTest.java (hunk 1 above)
Comment (lsyldliu) [reply flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/operators/aggregate/MiniBatchGroupAggFunctionTest.java:43]: Okay

File: flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/operators/aggregate/GroupAggFunctionTestBase.java (hunk 2 above)
Comment (lsyldliu) [reply flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/operators/aggregate/GroupAggFunctionTestBase.java]: @libenchao Thanks for your review, I have resolved your comments.

**Comment by lsyldliu** [comment]
@wuchong  should we continue this work because the performace degrade of incremental cleanup ?

**Comment by lsyldliu** [comment]
@flinkbot run azure

**Comment by lsyldliu** [comment]
@wuchong @libenchao some users have recently encountered state ttl problem when using minibatch aggregate, refer to [flink sql state size](https://www.mail-archive.com/user-zh@flink.apache.org/msg06493.html), [flink sql state clean](https://www.mail-archive.com/user-zh@flink.apache.org/msg06277.html), so I think we should add state ttl feature in 1.12 version, can you help review this pr again, tks!

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/aggregate/GroupAggFunction.java [hunk 3]
```diff
@@ -83,61 +85,57 @@
 	// stores the accumulators
 	private transient ValueState<RowData> accState = null;
 
+	private final StateTtlConfig ttlConfig;
```
Comment (wuchong) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/aggregate/GroupAggFunction.java:88]: nit: We can use `long stateRetentionTime` as the member variable to save the serialized state. I find you already declare stateRetentionTime in other classes, so would be better to keep consistent.

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/aggregate/GroupAggFunction.java [hunk 4]
```diff
@@ -83,61 +85,57 @@
...
-			long minRetentionTime,
-			long maxRetentionTime,
 			GeneratedAggsHandleFunction genAggsHandler,
 			GeneratedRecordEqualiser genRecordEqualiser,
 			LogicalType[] accTypes,
 			int indexOfCountStar,
-			boolean generateUpdateBefore) {
-		super(minRetentionTime, maxRetentionTime);
+			boolean generateUpdateBefore,
+			long minRetentionTime) {
```
Comment (wuchong) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/aggregate/GroupAggFunction.java:108]: nit: could you refactor the existing `minRetentionTime` to `stateRetentionTime`?

**Review by wuchong — COMMENTED** [review]
Sorry for the late response @lsyldliu . The changes looks good to me in general. I only left some minor comments. Could you rebase the master and trigger CI again?

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/aggregate/GroupAggFunction.java (hunk 4 above)
Comment (lsyldliu) [reply flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/aggregate/GroupAggFunction.java:108]: OK

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/aggregate/GroupAggFunction.java (hunk 3 above)
Comment (lsyldliu) [reply flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/aggregate/GroupAggFunction.java:88]: get

Comment (lsyldliu) [reply flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/aggregate/GroupAggFunction.java:88]: @wuchong sorry, I looked the code again, `ttlConfig` object also used in `processElement` method, so it must be global object. However, it only used in `open` method in other classes, to keep consistent, I think we can set ttlConfig as global object in other classes, what do you think？

Comment (wuchong) [reply flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/aggregate/GroupAggFunction.java:88]: I think we can simply use `stateRetentionTime > 0` to check whether ttl is enabled. We can also wrap that as a utility if needed.

**Comment by lsyldliu** [comment]
@flinkbot run azure

**Comment by lsyldliu** [comment]
@wuchong There are two tests failed, first is `RetractionITCase#testUniqueProcess`, error is _java.lang.AssertionError: expected:<List((false,1,1), (false,1,2), (false,12,1), (false,3,1), (false,6,1), (false,6,2), (true,1,1), (true,1,2), (true,1,3), (true,12,1), (true,18,1), (true,2,1), (true,3,1), (true,6,1), (true,6,1), (true,6,2), (true,8,1))> but was:<List((false,1,1), (false,1,1), (false,1,1), (false,1,1), (false,1,2), (false,1,2), (false,1,2), (false,1,2), (false,12,1), (false,3,1), (false,6,1), (false,6,2), (true,1,1), (true,1,1), (true,1,1), (true,1,1), (true,1,2), (true,1,2), (true,1,2), (true,1,2), (true,1,3), (true,12,1), (true,18,1), (true,2,1), (true,3,1), (true,6,1), (true,6,1), (true,6,2), (true,8,1))_
second is `AggregateITCase#testListAggWithRetraction`, error is **Caused by: java.lang.ClassCastException: org.apache.flink.runtime.state.ttl.TtlListState$$Lambda$1880/979931448 cannot be cast to java.util.List
**

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/aggregate/GroupAggFunction.java [hunk 5]
```diff
@@ -83,61 +85,58 @@
 	// stores the accumulators
 	private transient ValueState<RowData> accState = null;
 
+	private final long stateRetentionTime;
```
Comment (wuchong) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/aggregate/GroupAggFunction.java:88]: Move the this member variable around to those `final` variables.

**Comment by wuchong** [comment]
Regarding to `AggregateITCase#testListAggWithRetraction`, we should update `StateListView#remove` to not hard cast `getListState().get()` to `List`.  We can use the iterator for removing, and maybe we need to support `remove` for `TtlListState#IteratorWithCleanup`.

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/aggregate/GroupAggFunction.java [hunk 6]
```diff
@@ -180,7 +179,7 @@ public void processElement(RowData input, Context ctx, Collector<RowData> out) t
 
 			// if this was not the first row and we have to emit retractions
 			if (!firstRow) {
-				if (!stateCleaningEnabled && equaliser.equals(prevAggValue, newAggValue)) {
+				if (stateRetentionTime > 0 && equaliser.equals(prevAggValue, newAggValue)) {
```
Comment (wuchong) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/aggregate/GroupAggFunction.java:182]: I think this is the one causes `RetractionITCase#testUniqueProcess` failed. Here should be `stateRetentionTime <= 0` to indicate state cleaning is not enabled.

Comment (lsyldliu) [reply flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/aggregate/GroupAggFunction.java:182]: you are write

**Comment by lsyldliu** [comment]
> Regarding to `AggregateITCase#testListAggWithRetraction`, we should update `StateListView#remove` to not hard cast `getListState().get()` to `List`. We can use the iterator for removing, and maybe we need to support `remove` for `TtlListState#IteratorWithCleanup`.

for this problem, should we open a new jira issue talk about it?

**Review by wuchong — COMMENTED** [review]
Hi @lsyldliu , I just find that the current implementation doesn't cover the state cleanup for DataViews (i.g. ListView, MapView).  This will result in the state size increases unlimited when `count distinct` or `listagg` is used. We should pass the `StateTtlConfig` to the creation of StateDataView (e.g. `PerKeyStateDataViewStore`).

**Comment by lsyldliu** [comment]
> Hi @lsyldliu , I just find that the current implementation doesn't cover the state cleanup for DataViews (i.g. ListView, MapView). This will result in the state size increases unlimited when `count distinct` or `listagg` is used. We should pass the `StateTtlConfig` to the creation of StateDataView (e.g. `PerKeyStateDataViewStore`).

Hi, @wuchong , I have pass the `StateTtlConfig` to `PerKeyStateDataViewStore ` as constructing parameter, when create `StateDataView` we will judge `StateTtlConfig` whether enable state TTL, so we can clean the StateDataView state timely for `count distinct` or `listagg`.

**Comment by lsyldliu** [comment]
@wuchong I have updated the code, azure test passed, you can help review it again!

**Comment by lsyldliu** [comment]
CC @wuchong, we have tested it in vip internal demand.

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/dataview/PerKeyStateDataViewStore.java [hunk 7]
```diff
@@ -54,6 +61,9 @@ public PerKeyStateDataViewStore(RuntimeContext ctx) {
 			keySerializer,
 			valueSerializer);
 
+		if (stateTtlConfig.isEnabled()) {
+			mapStateDescriptor.enableTimeToLive(stateTtlConfig);
```
Comment (libenchao) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/dataview/PerKeyStateDataViewStore.java:65]: also need to make `nullStateDescriptor` aware of the TTL?

**Comment by wuchong** [comment]
I helped to rebase the PR and improve som codes, the modification includes:

1. Move the harness tests to the existing `GroupAggregateHarnessTest` using the code generated function by framework.
2. Enable TTL for the null state in MapView. Also cover the null state in the above test. 
3. Fix bug in `MiniBatchGroupAggFunction` to ignore retractions if state is cleaned up.
4. Improve some code.

**Comment by wuchong** [comment]
@leonardBang , could you help to review this? 

Appreciate if @lsyldliu can have a look at the changes too.

**Comment by lsyldliu** [comment]
> @leonardBang , could you help to review this?
> 
> Appreciate if @lsyldliu can have a look at the changes too.

Thanks a lot, LGTM

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/dataview/StateListView.java [hunk 8]
```diff
@@ -76,7 +77,14 @@ public void addAll(List<EE> list) throws Exception {
 
 	@Override
 	public boolean remove(EE value) throws Exception {
-		List<EE> list = (List<EE>) getListState().get();
+		// the getListState().get() not always returns List object
+		// copy values to ArrayList for removing
+		Iterator<EE> iterator = getListState().get().iterator();
```
Comment (leonardBang) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/dataview/StateListView.java:82]: `getListState().get()` may return null object, add a null check?

**Review by leonardBang — APPROVED** [review]
Thanks all for the contribution, LGTM, +1 to merge once the build passed

**Comment by wuchong** [comment]
Build is passed: https://dev.azure.com/apache-flink/apache-flink/_build/results?buildId=9847&view=results

File: flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/operators/aggregate/GroupAggFunctionTestBase.java (hunk 2 above)
Comment (uglymagic) [reply flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/operators/aggregate/GroupAggFunctionTestBase.java]: 1

**Pull request closed without merging**